src/Build/Instance/TaskRegistry.cs

```diff
@@ -1568,7 +1567,8 @@ private bool GetTaskFactory(TargetLoggingContext targetLoggingContext, ElementLo
                                         initialized = factory.Initialize(RegisteredName, ParameterGroupAndTaskBody.UsingTaskParameters, ParameterGroupAndTaskBody.InlineTaskXmlBody, taskFactoryLoggingHost);
 
                                         // TaskFactoryParameters will always be null unless specifically created to have runtime and architecture parameters.
-                                        if (initialized && TaskFactoryParameters != null)
+                                        // In case TaskHostFactory is explicitly requested, we will now have a parameter for that.
+                                        if (initialized && TaskFactoryParameters != null && !TaskFactoryParameters.TryGetValue(Constants.TaskHostExplicitlyRequested, out _))
```

Comment:
let's check here explicitly for the presence of architecture and runtime params
I suspect it's possible to have 3 of them set and it will cause a bug.